support/upsert/apply.go

```diff
@@ -0,0 +1,143 @@
+	loopDetector *updateLoopDetector
+}
+
+func NewApplyProvider(enableUpdateLoopDetector bool) ApplyProvider {
+	p := &applyProvider{}
+	if enableUpdateLoopDetector {
+		p.loopDetector = newUpdateLoopDetector()
+	}
+	return p
+}
+
+// ApplyManifest updates a resource from a yaml manifest configuration. The resource will be created if it doesn't exist yet.
+// This doesn't update status, use 'controllerutil.CreateOrPatch()' instead.
+func (p *applyProvider) ApplyManifest(ctx context.Context, c crclient.Client, obj crclient.Object) (controllerutil.OperationResult, error) {
+	existing := obj.DeepCopyObject().(crclient.Object) //nolint
+
+	if err := c.Get(ctx, crclient.ObjectKeyFromObject(obj), existing); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return controllerutil.OperationResultNone, err
+		}
+		if err := c.Create(ctx, obj); err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+
+		return controllerutil.OperationResultCreated, nil
+	}
+
+	result, err := p.update(ctx, c, obj, existing)
+	if err != nil || result == controllerutil.OperationResultNone {
+		return result, err
+	}
+
+	return controllerutil.OperationResultUpdated, nil
+}
+
+func (p *applyProvider) update(ctx context.Context, c crclient.Client, obj crclient.Object, existing crclient.Object) (controllerutil.OperationResult, error) {
+	key := crclient.ObjectKeyFromObject(obj)
+
+	switch existingTyped := existing.(type) {
+	case *corev1.ServiceAccount:
+		preserveServiceAccountPullSecrets(existingTyped, obj.(*corev1.ServiceAccount))
+	}
+	preserveOriginalMetadata(existing, obj)
+
+	current, err := toUnstructured(existing)
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+	modified, err := toUnstructured(obj)
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	// DeepDerivative ignores unset fields in 'modified' (empty/nil arrays, empty strings, etc.)
+	if equality.Semantic.DeepDerivative(modified, current) {
+		if p.loopDetector != nil {
+			p.loopDetector.recordNoOpUpdate(obj, key)
+		}
+		return controllerutil.OperationResultNone, nil
+	}
+
+	if p.loopDetector != nil {
+		p.loopDetector.recordActualUpdate(existing, obj, key)
+	}
+
+	if err := c.Update(ctx, obj); err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	return controllerutil.OperationResultUpdated, nil
+}
+
+func preserveOriginalMetadata(original, mutated crclient.Object) {
+	labels := original.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	maps.Copy(labels, mutated.GetLabels())
+	mutated.SetLabels(labels)
+
+	annotations := original.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	maps.Copy(annotations, mutated.GetAnnotations())
+	mutated.SetAnnotations(annotations)
+
+	finalizers := sets.New(original.GetFinalizers()...).Insert(mutated.GetFinalizers()...)
+	mutated.SetFinalizers(sets.List(finalizers))
+
+	mutated.SetResourceVersion(original.GetResourceVersion())
```

Comment:
why is the ResourceVersion not ignored?

Reply:
for future reference: This needs to be set and passed back by clients for concurrency checks
ref: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency

Reply:
fwiw how you implement passing that back might be tied to your patching strategy, e.g. I don't think we'd set it if using SSA

Reply:
true, I have been looking at the patch code from controller-util and they set this field only when optimistLock is true
see:
https://github.com/kubernetes-sigs/controller-runtime/blob/5dfe3216fb7fd7f5afb61d6d0f8956c7bec8df62/pkg/client/patch.go#L104-L115